tests/evaluate/test_evaluation.py

```diff
@@ -1567,3 +1567,163 @@ def test_evaluate_with_static_dataset_error_handling_pandas_dataset():
                 model_type="regressor",
                 predictions="model_output",
             )
+
+
+_DUMMY_CHAT_RESPONSE = {
+    "id": "1",
+    "object": "text_completion",
+    "created": "2021-10-01T00:00:00.000000Z",
+    "model": "gpt-3.5-turbo",
+    "choices": [
+        {
+            "index": 0,
+            "message": {
+                "content": "This is a response",
+                "role": "assistant",
+            },
+            "finish_reason": "length",
+        }
+    ],
+    "usage": {
+        "prompt_tokens": 1,
+        "completion_tokens": 1,
+        "total_tokens": 2,
+    },
+}
+
+_DUMMY_COMPLETION_RESPONSE = {
+    "id": "1",
+    "object": "text_completion",
+    "created": "2021-10-01T00:00:00.000000Z",
+    "model": "gpt-3.5-turbo",
+    "choices": [{"index": 0, "text": "This is a response", "finish_reason": "length"}],
+    "usage": {
+        "prompt_tokens": 1,
+        "completion_tokens": 1,
+        "total_tokens": 2,
+    },
+}
+
+
+@mock.patch("mlflow.models.evaluation.llm_utils.requests.post")
```

Comment:
Can we use `with mock.patch` instead to minize the scope of monkey-patching?


```python
def test_foo():
    a()
    with mock.patch(...):
        b()
    c()
```

is better than:

```
@mock.patch(...)
def test_foo():
    a()
    b()
    c()
```

Reply:
Hmm but annotation also has pros of making the code flat and telling what modules are mocked for the test case? Also it helps catch unexpected call to the module.

Reply:
@B-Step62 

> making the code flat (and prep -> run -> validate structure)

Can we achieve the same thing using `with mock.patch()`?

```
with mock.patch("foo.bar", return_value=foo) as mock_foo:  # prep
    do_foo()  # run

    mock_foo.xxx_yyy()  # validate
```

I'm not a big fan of the decorator `mock.patch` because I need to guess which lines require mocks and want to minimize the scope of monkey-patching to avoid false positives.

Reply:
Yea I'm fine with both, just wanted to avoid nesting `with` statement like this. But it's not too ugly in this case after all.

```
with mock.patch(...) as mock_deploy_client:
     # do some mock setup
     with mlflow.start_run():
        ...
```

Speaking about the annotator itself, my position is that there is no clear winner between two options (e.g. false-positive can happen in the other way - let's say we want to mock and count calls to tracking server, then it may call to local server outside the content). The annotator works better for structuring the test method evenly, particularly if you have many test cases with multiple mocks, it much easier to read. But specifically to this case I don't see much difference anyway.

Reply:
@B-Step62 Fair enough. Let's use the decorator :)